ddtrace/contrib/vertica/patch.py

```diff
@@ -0,0 +1,246 @@
+well. wrapt will hook into the module import system for the patches that we
+provide.
+"""
+
+_PATCHED = False
+
+def copy_before(instance, span, conf, *args, **kwargs):
+    span.set_tag("query", args[0])
+
+def execute_before(instance, span, conf, *args, **kwargs):
+    span.set_tag("query", args[0])
+
+def execute_after(result, instance, span, conf, *args, **kwargs):
+    span.set_metric(dbx.ROWCOUNT, instance.rowcount)
+
+def fetch_after(result, instance, span, conf, *args, **kwargs):
+    span.set_metric(dbx.ROWCOUNT, instance.rowcount)
+
+def cursor_after(cursor, instance, span, conf, *args, **kwargs):
+    tags = {}
+    tags[net.TARGET_HOST] = instance.options["host"]
+    tags[net.TARGET_PORT] = instance.options["port"]
+    if "user" in instance.options:
+        tags[dbx.USER] = instance.options["user"]
+    if "database" in instance.options:
+        tags[dbx.NAME] = instance.options["database"]
+
+    pin = Pin(
+        service=config.vertica["service_name"],
+        app=APP,
+        app_type=AppTypes.db,
+        tags=tags,
+        _config=config.vertica["patch"]["vertica_python.vertica.cursor.Cursor"]
+    )
+    pin.onto(cursor)
+
+
+# tracing configuration
+config._add(
+    "vertica",
+    {
+        "service_name": "vertica",
+        "app": "vertica",
+        "app_type": "db",
+        "patch": {
+            "vertica_python.vertica.connection.Connection": {
+                "routines": {
+                    "cursor": {
+                        "trace_enabled": False,
+                        "on_after": cursor_after,
+                    }
+                },
+            },
+            "vertica_python.vertica.cursor.Cursor": {
+                "routines": {
+                    "execute": {
+                        "operation_name": "vertica.query",
+                        "span_type": "vertica",
+                        "on_before": execute_before,
+                        "on_after": execute_after,
+                    },
+                    "copy": {
+                        "operation_name": "vertica.copy",
+                        "span_type": "vertica",
+                        "on_before": copy_before,
+                    },
+                    "fetchone": {
+                        "operation_name": "vertica.fetchone",
+                        "span_type": "vertica",
+                        "on_after": fetch_after,
+                    },
+                    "fetchall": {
+                        "operation_name": "vertica.fetchall",
+                        "span_type": "vertica",
+                        "on_after": fetch_after,
+                    },
+                    "nextset": {
+                        "operation_name": "vertica.nextset",
+                        "span_type": "vertica",
+                        "on_after": fetch_after,
+                    },
+                },
+            },
+        },
+    },
+)
+
+def patch():
+    global _PATCHED
```

Comment:
I don't like this mechanism, but since we cannot import the library, we cannot use the same mechanism we do in other integrations.

Reply:
we don't need a mutex lock here for modifying `_PATCHED` do we?

Reply:
Don't think so. We don't use any kind of locking mechanism for our other integrations and patching/unpatching shouldn't occur in multithreaded environments.

Reply:
:+1:

Reply:
@labbati makes a good point.

If you did something like:


```python
import ddtrace; ddtrace.patch_all(vertica=True)

import vertica

# Reload
import importlib
importlib.reload_module(vertica)

# This is a fun one
import sys
del sys.modules['vertica']
import vertica
```

Or basically anything that causes the `vertica` module to get re-evaluated and loaded into `sys.modules` then we will no longer be patched.


This is probably realistically not an issue here, but a good big picture conversation for how we should be doing this across all of our integrations.

Reply:
Yes, you both raise valid concerns. We could assign an attribute when we add a `Pin` to the class on `__init__` and check for the existence of this attribute (similarly to how our other integrations work). This should cover us for now but we should look into alternative solutions later on.

Reply:
On second thought, we can't really take this approach since we still need to import the module at some point. I'm leaning towards leaving it as is for now. We'll have to consider this when circling back to the new config api.

Reply:
Your call, I think this is good to merge as-is. If you do try to make a change, time box it, and if it gives you any hassle then lets go with the current approach.

Given we want to merge this soon we want to limit the chances of unknowns making it into a release... this might not be an ideal approach, but at least we know how it'll work.